batch/batch/driver/scheduler.py

```diff
@@ -111,33 +120,143 @@ def allocate_cores(user, mark):
     async def bump_loop(self):
         while True:
             self.scheduler_state_changed.set()
-            self.cancel_state_changed.set()
+            self.cancel_ready_state_changed.set()
+            self.cancel_running_state_changed.set()
             await asyncio.sleep(60)
 
-    async def cancel_loop_body(self):
-        user_records = self.db.select_and_fetchall(
+    async def cancel_cancelled_ready_jobs_loop_body(self):
+        records = self.db.select_and_fetchall(
             '''
-SELECT user
-FROM user_resources
-GROUP BY user
-HAVING COALESCE(SUM(running_cores_mcpu), 0) > 0;
+SELECT user, n_cancelled_ready_jobs
+FROM (SELECT user,
+    CAST(COALESCE(SUM(n_cancelled_ready_jobs), 0) AS SIGNED) AS n_cancelled_ready_jobs
+  FROM user_resources
+  GROUP BY user) AS t
+WHERE n_cancelled_ready_jobs > 0;
 ''')
+        user_n_cancelled_ready_jobs = {
+            record['user']: record['n_cancelled_ready_jobs'] async for record in records
+        }
+
+        total = sum(user_n_cancelled_ready_jobs.values())
+        user_share = {
+            user: max(int(1000 * user_n_jobs / total + 0.5), 20)
+            for user, user_n_jobs in user_n_cancelled_ready_jobs.items()
+        }
+
+        async def user_cancelled_ready_jobs(user, remaining):
+            async for batch in self.db.select_and_fetchall(
+                    '''
+SELECT id, cancelled
+FROM batches
+WHERE user = %s AND `state` = 'running';
+''',
+                    (user,)):
+                if batch['cancelled']:
+                    async for record in self.db.select_and_fetchall(
+                            '''
+SELECT jobs.job_id
+FROM jobs
+WHERE batch_id = %s AND always_run = 0 AND state = 'Ready'
+LIMIT %s;
+''',
+                            (batch['id'], remaining.value)):
+                        record['batch_id'] = batch['id']
+                        yield record
+                else:
+                    async for record in self.db.select_and_fetchall(
+                            '''
+SELECT jobs.job_id
+FROM jobs
+WHERE batch_id = %s AND always_run = 0 AND state = 'Ready' AND cancelled = 1
```

Comment:
No, this second part is for a batch which is not cancelled (so `batch['cancelled']` is false) but the job itself is cancelled (because a parent failed).  That's why we check cancelled here.